### PR TITLE
HPCC-14844 Cache database connections when using Embedded SQL

### PR DESCRIPTION
**Pull request by richardkchapman**
Cache multiple conenctions from same thread.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

**Comment by ghalliday** [comment]
@richardkchapman 

Should this also be calling mysql_thread_end() when the thread is exiting?  See https://dev.mysql.com/doc/refman/5.7/en/c-api-threaded-clients.html.

See also notes on mysql_init() thread safety https://dev.mysql.com/doc/refman/5.7/en/mysql-init.html

**Comment by richardkchapman** [comment]
The docs are not especially clear about whether the thread_end call is needed only if you call other mysql functions on a thread before you call mysql_init... Either way I think any issues related to MySQL multi-thread behaviour would pre-exist this PR...

**Comment by richardkchapman** [comment]
@ghalliday Added some thread-init calls

File: plugins/mysql/mysqlembed.cpp [hunk 1]
```diff
@@ -956,6 +956,73 @@ class MySQLRowStream : public CInterfaceOf<IRowStream>
...
+static bool mysqlInitialized = false;
+static __thread bool mysqlThreadInitialized = false;
+static CriticalSection initCrit;
+
+static void terminateMySqlThread()
+{
+    clearCache();
+    mysql_thread_end();
+    mysqlThreadInitialized = false;  // In case it was a threadpool thread...
+    threadHookChain = NULL;
```
Comment (ghalliday) [inline plugins/mysql/mysqlembed.cpp]: need to call threadHookChain() if non null.

**Pull request merged**